R/layer_residual_quantiles.R

```diff
@@ -34,46 +36,79 @@
 #' wf1 <- wf %>% add_frosting(f)
 #'
 #' p <- predict(wf1, latest)
-#' p
+#'
+#' f2 <- frosting() %>%
+#'   layer_predict() %>%
+#'   layer_residual_quantiles(probs = c(0.3, 0.7), by_key = "geo_value") %>%
+#'   layer_naomit(.pred)
+#' wf2 <- wf %>% add_frosting(f2)
+#'
+#' p2 <- predict(wf2, latest)
 layer_residual_quantiles <- function(frosting, ...,
                                      probs = c(0.0275, 0.975),
                                      symmetrize = TRUE,
+                                     by_key = character(0L),
                                      name = ".pred_distn",
                                      .flag = TRUE,
                                      id = rand_id("residual_quantiles")) {
   rlang::check_dots_empty()
+  arg_is_scalar(symmetrize, .flag)
   arg_is_chr_scalar(name, id)
+  arg_is_chr(by_key, allow_null = TRUE)
   arg_is_probabilities(probs)
-  arg_is_lgl(symmetrize)
+  arg_is_lgl(symmetrize, .flag)
   add_layer(
     frosting,
     layer_residual_quantiles_new(
       probs = probs,
       symmetrize = symmetrize,
+      by_key = by_key,
       name = name,
       id = id
     ),
     flag = .flag
   )
 }
 
-layer_residual_quantiles_new <- function(probs, symmetrize, name, id) {
+layer_residual_quantiles_new <- function(probs, symmetrize, by_key, name, id) {
   layer("residual_quantiles", probs = probs, symmetrize = symmetrize,
-        name = name, id = id)
+        by_key = by_key, name = name, id = id)
 }
 
 #' @export
 slather.layer_residual_quantiles <-
   function(object, components, the_fit, the_recipe, ...) {
     if (is.null(object$probs)) return(components)
 
+
     s <- ifelse(object$symmetrize, -1, NA)
-    r <- grab_residuals(the_fit, components)
-    q <- quantile(c(r, s * r), probs = object$probs, na.rm = TRUE)
+    r <- dplyr::bind_cols(
+      r = grab_residuals(the_fit, components),
+      geo_value = components$mold$extras$roles$geo_value,
+      components$mold$extras$roles$key)
```

Comment:
What is the column name for this? I know there are `r` and `geo_value` column names.

Reply:
`components$mold$extras$roles$key` can be 2 things.
1. It will be `NULL` if there are no additional keys. In this case, `r` will contain 2 columns `r` and `geo_value`
2. It will be a data.frame with one column for each key. Then this data frame will be bound as is to the other two columns, preserving the column names.